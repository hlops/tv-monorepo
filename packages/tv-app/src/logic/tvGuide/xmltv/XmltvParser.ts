import _ from 'lodash';
import moment from 'moment';
import {QualifiedTag, Tag} from "sax";
import * as sax from 'sax';
import { SAXStream } from 'sax';
import stream from 'stream';
import Readable from 'stream';
import zlib = require('zlib');

type Node = Tag;

export interface Programme {
	start: Date;
	stop: Date;
	title: string;
	categories?: [];
	desc?: string;
}

export class XmltvParser extends stream.Writable {
	private _currentNode?: Node;
	private _parentNodes: Node[] = [];
	private _xmlParser: SAXStream;

	constructor() {
		super();

		const channels: _.Dictionary<string> = {};

		this._xmlParser = sax.createStream(true, {
			trim: true,
			position: false
		});

		this.on('finish', () => {
			this._xmlParser.end();
		});

		this._xmlParser.on('end', () => {
			this.emit('end');
		});

		this._xmlParser.on('error', error => {
			this.emit('error', error);
		});

		this._xmlParser.on('opentag', (node: Node) => {
			this._currentNode = node;
			if (!_.get(_.last(this._parentNodes), 'isSelfClosing')) {
				this._parentNodes.push(node);
			}
		});

		this._xmlParser.on('closetag', (tagName: string) => {
			if (tagName === 'programme') {
				let node = this._parentNodes[this._parentNodes.length - 1];
				this.emit('programme', channels[node.attributes.channel], {
					start: XmltvParser.parseDate(node.attributes.start),
					stop: XmltvParser.parseDate(node.attributes.stop),
					title: node.attributes.title,
					categories: node.attributes.categories,
					desc: node.attributes.desc
				});
			}
			this._parentNodes.pop();
		});

		this._xmlParser.on('text', text => {
            const node = this._parentNodes[this._parentNodes.length - 2];
			if (node && this._currentNode) {
				switch (this._currentNode.name) {
					case 'display-name': {
						if (node.name === 'channel') {
							channels[node.attributes.id] = text;
						}
						break;
					}
					case 'category': {
						if (node.name === 'programme') {
							if (!node.attributes.categories) {
								_.set(node.attributes, 'categories',  []);
							}
							const categories = node.attributes.categories as unknown;
                            (categories as string[]).push(text);
						}
						break;
					}
					case 'title':
					case 'desc': {
						if (node && node.name === 'programme') {
							node.attributes[this._currentNode.name] = text;
						}
					}
				}

				const nodeName = this._currentNode.name;
				if (nodeName === 'display-name') {
				} else if (nodeName === 'display-name') {
				} else if (['title', 'desc'].indexOf(nodeName) >= 0) {
				}
			}
		});
	}

	static parseDate(date: string): Date {
		const parsed = moment(date, 'YYYYMMDDHHmmss Z', true);
		return parsed.toDate();
	}

	_write(chunk: any, encoding: string, callback: Function): void {
		this._xmlParser.write(chunk, encoding);
		callback();
	}

	public parseGzip(input: Readable): Promise<_.Dictionary<Programme[]>> {
		return this.parse(input.pipe(zlib.createGunzip()));
	}

	public parse(input: Readable): Promise<_.Dictionary<Programme[]>> {
		const parser = new XmltvParser();
		input.pipe(parser);
		const channels: _.Dictionary<Programme[]> = {};

		return new Promise((resolve, reject) => {
			parser.on('programme', (channel: string, programme: Programme) => {
				if (!channels[channel]) {
					channels[channel] = [];
				}
				channels[channel].push(programme);
			});

			parser.on('end', () => {
				resolve(channels);
			});

			parser.on('error', error => {
				reject(error);
			});
		});
	}
}
