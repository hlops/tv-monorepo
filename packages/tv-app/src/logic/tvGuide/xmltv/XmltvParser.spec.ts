import fs from 'fs';
import _ from "lodash";
import { XmltvParser } from './XmltvParser';

describe('XmltvParser', () => {
	it('parse() can parse xmltv file from resources', done => {
		new XmltvParser()
			.parse(fs.createReadStream('./samples/xmltv-small.xml', 'utf8'))
			.then(channels => {
                expect(channels).toMatchSnapshot();
				done();
			});
	});
});
