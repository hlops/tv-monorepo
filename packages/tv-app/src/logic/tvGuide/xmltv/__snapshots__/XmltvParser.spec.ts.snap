// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XmltvParser parse() can parse xmltv file from resources 1`] = `
Object {
  "CBS Affiliate": Array [
    Object {
      "categories": Array [
        "Talk",
        "News",
        "Series",
      ],
      "desc": "Republican candidate John McCain; premiere of the film \\"The Dark Knight.\\"",
      "start": 2008-07-15T12:00:00.000Z,
      "stop": 2008-07-15T14:00:00.000Z,
      "title": "The Early Show",
    },
    Object {
      "categories": Array [
        "Talk",
        "Series",
      ],
      "desc": "Actresses Kim Raver, Brooke Shields and Lindsay Price (\\"Lipstick Jungle\\"); women in their 40s tell why they got breast implants; a 30-minute meal.",
      "start": 2008-07-15T14:00:00.000Z,
      "stop": 2008-07-15T15:00:00.000Z,
      "title": "Rachael Ray",
    },
    Object {
      "categories": Array [
        "Game show",
        "Series",
      ],
      "desc": "Contestants bid for prizes then compete for fabulous showcases.",
      "start": 2008-07-15T15:00:00.000Z,
      "stop": 2008-07-15T16:00:00.000Z,
      "title": "The Price Is Right",
    },
    Object {
      "categories": Array [
        "Game show",
        "Series",
      ],
      "desc": undefined,
      "start": 2008-07-15T16:00:00.000Z,
      "stop": 2008-07-15T16:30:00.000Z,
      "title": "Jeopardy!",
    },
    Object {
      "categories": Array [
        "Soap",
        "Series",
      ],
      "desc": "Jeff thinks Kyon stole the face cream; Nikki asks Nick to give David a chance; Amber begs Adrian to go to Australia.",
      "start": 2008-07-15T16:30:00.000Z,
      "stop": 2008-07-15T17:30:00.000Z,
      "title": "The Young and the Restless",
    },
  ],
  "PBS Affiliate": Array [
    Object {
      "categories": Array [
        "Newsmagazine",
        "Interview",
        "Public affairs",
        "Series",
      ],
      "desc": "Jordan's Queen Rania has made job creation a priority to help curb the staggering unemployment rates among youths in the Middle East.",
      "start": 2008-07-15T06:30:00.000Z,
      "stop": 2008-07-15T07:00:00.000Z,
      "title": "NOW on PBS",
    },
    Object {
      "categories": Array [
        "Anthology",
        "Mystery",
        "Series",
      ],
      "desc": "Foyle investigates an explosion at a munitions factory, which he comes to believe may have been premeditated.",
      "start": 2008-07-15T07:00:00.000Z,
      "stop": 2008-07-15T08:30:00.000Z,
      "title": "Mystery!",
    },
    Object {
      "categories": Array [
        "Anthology",
        "Mystery",
        "Series",
      ],
      "desc": "The murder of a prominent scientist may have been due to a gambling debt.",
      "start": 2008-07-15T08:30:00.000Z,
      "stop": 2008-07-15T10:00:00.000Z,
      "title": "Mystery!",
    },
    Object {
      "categories": Array [
        "News",
        "Series",
      ],
      "desc": "International issues.",
      "start": 2008-07-15T10:00:00.000Z,
      "stop": 2008-07-15T10:30:00.000Z,
      "title": "BBC World News",
    },
    Object {
      "categories": Array [
        "Exercise",
        "Series",
      ],
      "desc": undefined,
      "start": 2008-07-15T10:30:00.000Z,
      "stop": 2008-07-15T11:00:00.000Z,
      "title": "Sit and Be Fit",
    },
  ],
}
`;
